libs/custom-scanner/src/protocol.ts

```diff
@@ -0,0 +1,1096 @@
+
+/**
+ * Gets the release version of the specified type.
+ */
+export async function getReleaseVersion(
+  channel: DuplexChannel,
+  releaseType: ReleaseType
+): Promise<Result<string, ErrorCode>> {
+  debug(
+    'getReleaseVersion releaseType=%s (%x)',
+    ReleaseType[releaseType],
+    releaseType
+  );
+  const result = await sendRequestAndReadResponse(
+    channel,
+    ReleaseVersionRequest,
+    { releaseType },
+    MAX_GET_RELEASE_VERSION_RESPONSE_LENGTH,
+    'data'
+  );
+
+  return mapResult(result, ({ data }) => data);
+}
+
+const MAX_GET_STATUS_INTERNAL_RESPONSE_LENGTH = 0x30;
+
+/**
+ * Gets the internal status of the scanner.
+ */
+export async function getStatusInternal(
+  channel: DuplexChannel,
+  jobId: number
+): Promise<Result<StatusInternalMessage, ErrorCode>> {
+  debug('getStatusInternal jobId=%x', jobId);
+  const result = await sendRequestAndReadResponse(
+    channel,
+    StatusInternalRequest,
+    { jobId },
+    MAX_GET_STATUS_INTERNAL_RESPONSE_LENGTH,
+    'other'
+  );
+
+  return mapResult(result, ({ buffer }) => {
+    return mapCoderError(StatusInternalMessage.decode(buffer));
+  });
+}
+
+/**
+ * Create a new job.
+ */
+export async function createJob(
+  channel: DuplexChannel
+): Promise<Result<number, ErrorCode>> {
+  debug('createJob');
+  const result = await sendRequestAndReadResponse(
+    channel,
+    JobCreateRequest,
+    undefined,
+    DEFAULT_MAX_READ_LENGTH,
+    'ack'
+  );
+
+  return mapResult(result, ({ jobId }) => jobId);
+}
+
+/**
+ * End a job.
+ */
+export async function endJob(
+  channel: DuplexChannel,
+  jobId: number
+): Promise<Result<void, ErrorCode>> {
+  debug('endJob jobId=%x', jobId);
+  const result = await sendRequestAndReadResponse(
+    channel,
+    JobEndRequest,
+    { jobId },
+    DEFAULT_MAX_READ_LENGTH,
+    'ack'
+  );
+
+  // drop the `jobId` from the response
+  return mapResult(result, () => undefined);
+}
+
+/**
+ * Move the form.
+ */
+export async function formMove(
+  channel: DuplexChannel,
+  jobId: number,
+  movement: FormMovement
+): Promise<Result<void, ErrorCode>> {
+  debug(
+    'formMove jobId=%x movement=%s (%x)',
+    jobId,
+    FormMovement[movement],
+    movement
+  );
+  const result = await sendRequestAndReadResponse(
+    channel,
+    FormMovementRequest,
+    { jobId, movement },
+    DEFAULT_MAX_READ_LENGTH,
+    'ack'
+  );
+
+  // drop the `jobId` from the response
+  return mapResult(result, () => undefined);
+}
+
+/**
+ * Set the scan parameters.
+ */
+export async function setScanParameters(
+  channel: DuplexChannel,
+  jobId: Uint8,
+  scanParameters: SetScanParametersRequestData
+): Promise<Result<void, ErrorCode>> {
+  debug('setScanParameters');
+  const setScanParametersResult = await sendRequest(
+    channel,
+    SetScanParametersRequest,
+    { jobId }
+  );
+
+  if (setScanParametersResult.isErr()) {
+    return setScanParametersResult;
+  }
+
+  const setScanParametersDataResult = await sendRequest(
```

Comment:
Because we send _two_ messages before expecting a response. The first is the "we're about to set scan parameters" command and the second is "here are the scan parameters". I don't know why they modeled it that way.

Reply:
Makes sense. Could the second message use `sendRequestAndReadResponse`? Just noticed some duplicated logic in the response handling

Reply:
Changed the second one to use `sendRequestAndReadResponse` and also made `sendRequestAndReadResponse` use `sendRequest`.